### ISSUE TITLE
New Action: insertOrUpdate

### ISSUE THREAD

**Issue by iNaD** [vuex-orm/vuex-orm#73]
Today I stumbled upon a common use case which currently isn't covered by vuex-orm.

## Idea

Add an action `insertOrUpdate` (couldn't think of a better name) which will insert new records into the collection or update records already existing.

```js
// Initial State
{
    entities: {
        users: {
            data: {
                '1': { id: 1, name: 'John', roles: [3] }
            }
        }
    }
}

// `insertOrUpdate` is going to add new records and update existing data (see `update`).
store.dispatch('entities/users/insertOrUpdate', {
    data: [
        { id: 1, name: 'Peter' },
        { id: 2, name: 'Hank' },
    ]
})
  
// State after `insertOrUpdate`.
{
    entities: {
        users: {
            data: {
                '1': { id: 1, name: 'Peter', roles: [3] },
                '2': { id: 2, name: 'Hank' }
            }
        }
    }
}
```

## Use case

In my case relationships aren't always returned by the API. You can specify which relationships should be embedded in the response to reduce data transfer and queries on the database.

E.g. when loading the current user, I will fetch his roles, but for listing all users this isn't needed, so I can fetch them when loading the detail view. But currently when loading all users without roles, all existing users will be overwritten with the provided data.

## Current workaround

My current workaround is looking like the following and I think it would be helpful for many users to include a similar solution in vuex-orm.

```js
const users; // Retrieved from some API
const promises = users.map(user => {
    if (this.$store.getters['entities/users/find'](user.id) !== null) {
        return this.$store.dispatch('entities/users/update', {
            where: user.id,
            data: user
        });
    }
    return this.$store.dispatch('entities/users/insert', {
        data: user
    });
});

Promise.all(promises).then(() => {
    // All users have been inserted or update, now we can do anything we want
});
```

## Final words

What are you thinking about this idea? Is it a good approach to extend vuex-orm or should I just use a self-written helper? If you like this idea, I could try to create a PR implementing this feature.

**Comment by kiaking** [comment]
Hi! Thank you for the request! I think this is a very good feature to have. I think the method name `insertOrUpdate` is good too.

> Is it a good approach to extend vuex-orm or should I just use a self-written helper?

I think it is good for Vuex ORM to have this feature 👍 

> If you like this idea, I could try to create a PR implementing this feature.

That would be super awesome! However, the implementation should be in Repo class. So, if you're not so familiar with the Repo class, I could implement this. What do you think?

**Comment by iNaD** [comment]
Thank you for the fast response. I will have a look at the Repo and how it processes the create and update actions. If I get stuck I will write you 😉

**Comment by timoschwarzer** [comment]
Can we extend the update mutation with a property `insertMissing`?

**Comment by kiaking** [comment]
@iNaD Great! Thanks. I think you can follow `repo.insert` method to see what it is doing (it actually calling Query class to perform non relational things such as fetching record or modifying states). Maybe I should document this but the role is;

**Repo** – Responsible for handling data normalization and relationals. Repo will not mutate or reference state directly but do anything related to relation such as getting related models or normalize data before passing data to the Query.

**Query** – Responsible for actually mutating or referencing Vuex State. So ideally Query will also have `insertOrUpdate` method and Repo will just call this method in this case, I guess.

@timoschwarzer That could be a nice idea since we have `create` option in insert method. API feels consistent. I think it's nice to have `insertOrUpdate` method too though. I like the idea of having shorthand or alias methods so after implementing `insertOrUpdate` method, we could use it through `insertMissing` option as well.

Will that option take boolean? or would it take relation name so that users can control which relation get's which?

**Comment by timoschwarzer** [comment]
Maybe `insertMissing: true` → Insert all missing data, `insertMissing: ['foo', 'bar']` → Insert boo and bar when updating.

**Comment by kiaking** [comment]
Hah, that's nice idea. However, when I think about it, I don't think it's possible to pass relational (nested) data to the `update` action...

We should first update that and then add that option I guess.

**Comment by kiaking** [comment]
Thanks to @iNaD, this feature is now implemented in Vuex ORM 0.19.0 🎉 
https://github.com/vuex-orm/vuex-orm/releases/tag/v0.19.0
